### ISSUE TITLE
Comments on the shiny persistent storage draft

### ISSUE THREAD

**Issue by jennybc** [daattali/daattali.github.io#4]
First, it's great!

Stream of consciousness as I read and try things:

My first attempt to add a row via the demo app using googlesheets resulted in an error (canno make the connection or some such) but second attempt worked. I'm not familiar w/ that error from googlesheets work FWIW.

When you first present code for the basic app, maybe clarify this is ... what, a combiination of the stuff that ultimately goes in `server.R` and `ui.R`?

I really like how you isolated the bits that change in `saveData()` and `loadData()`.

> The googlesheets package uses the same API that rdrop2 uses, and thus also needs to get authenticated in a similar fashion, such as by copying a valid .httr-oauth file to your Shiny directory.

I find this reference to a shared API confusing. Maybe rephrase as ... googlesheets and rDrop2 use a similar approach to authentication. I also wouldn't really advise doing this via `.httr-oauth`, but I suppose you can and apparently that's working for your demo app. It is the simplest. But I think it's better to explicitly save a token and load it. But I am explaining that in demo apps within googlesheets.

For `gs_add_row()` to work right now, it's needs a header row AND one row of data. I'm going to work on that FYI. So you don't necessarily need to point that out.

As I write this, Jeroen Ooms is presenting on mongolite. Did you think about using that package for your mongo example?

Again, nice work!

**Comment by daattali** [comment]
Fixed the googlesheets error and addressed the other issues as well, thanks.

Did not know about mongolite, it must be very new cuz I thought I did an exhaustive search before. The package I'm using seems to work just fine and still supported by original author, so unless you have a great reason to use the newer one, I'll leave it as is.

httr-oauth vs saving/loading token: do you know if there's an official Hadley opinion on that, or has there been any major discussion anywhere on which method is better to use with Shiny?

**Comment by jennybc** [comment]
Re: mongolite. I don't have a great reason, no. But if it goes down like with JSON and `jsonlite`, Jeroen's package will become the default. I mention it only so you could just go ahead and use it now. The post might stand the test of time better that way?

Re: the token. I don't think Hadley has a position. I think the very little experience we have with this with `googlesheets` + Shiny is as much as anyone has? `.httr-oauth` is a file that can potentially store tokens for multiple combinations of endpoint, app, etc. A hash of that identifying info is used as a key to store the token in `.httr-oauth`. When you want to hard-wire an app to access a particular Sheet with specific credentials, it probably makes sense to store the token as an isolated object in an `.rds` file and load it explicitly, versus relying on auto-loading from the cache in a general purpose `.httr-oauth` file. But honestly it probably doesn't matter that much. I'm doing it the other way in our demo apps, but I can understand why you want to keep it very simple.

**Comment by daattali** [comment]
**mongolite:** fair point, it might be better in the long run. It didn't just work out of the box for me because I was using a remote server, I had to help Jeroen debug, so I'm going to stick with the older & more established package

**auth:** I now see the use of rds when wanting to allow an app access to only a specific sheet. I'd rather just give the basic general solution, and anyone who actually ends up using googlesheets for real will hopefully read your README for more specifics. Maybe I'll just throw in another sentence encouraging the user to read the readme of the package being used rather than just relying only on the code I provide.
